vault/token_store.go

```diff
@@ -1213,57 +1213,118 @@ func (ts *TokenStore) handleCreateCommon(
+		if len(data.Policies) > 0 {
+			finalPolicies = policyutil.SanitizePolicies(data.Policies, localAddDefault)
 		}
 
-	case role != nil && len(role.DisallowedPolicies) > 0:
-		if len(data.Policies) == 0 {
-			data.Policies = parent.Policies
+		var sanitizedRolePolicies []string
+
+		// First check allowed policies; if policies are specified they will be
+		// checked, otherwise if an allowed set exists that will be the set
+		// that is used
+		if len(role.AllowedPolicies) > 0 {
+			// Note that if "default" is already in allowed, and also in
+			// disallowed, this will still result in an error later since this
+			// doesn't strip out default
+			sanitizedRolePolicies = policyutil.SanitizePolicies(role.AllowedPolicies, localAddDefault)
+
+			if len(finalPolicies) == 0 {
+				finalPolicies = sanitizedRolePolicies
+			} else {
+				if !strutil.StrListSubset(sanitizedRolePolicies, finalPolicies) {
+					return logical.ErrorResponse(fmt.Sprintf("token policies (%v) must be subset of the role's allowed policies (%v)", finalPolicies, sanitizedRolePolicies)), logical.ErrInvalidRequest
+				}
+			}
+		} else {
+			// Check against parent policies, or assign parent policies. As
+			// this is a role, add default unless explicitly disabled.
+			if len(finalPolicies) == 0 {
+				finalPolicies = policyutil.SanitizePolicies(parent.Policies, localAddDefault)
+			} else {
+				// If we added default based on the fact that this is using a
+				// role, we need to add it here too to ensure that the subset
+				// matching works.
+				sanitizedParentPolicies := policyutil.SanitizePolicies(parent.Policies, localAddDefault)
+				if !strutil.StrListSubset(sanitizedParentPolicies, finalPolicies) {
+					return logical.ErrorResponse("child policies must be subset of parent when role contains no allowed_policies"), logical.ErrInvalidRequest
+				}
+			}
 		}
-		sanitizedInputPolicies := policyutil.SanitizePolicies(data.Policies, true)
 
-		// Do not voluntarily add 'default' to the list of items to check on
-		sanitizedRolePolicies := policyutil.SanitizePolicies(role.DisallowedPolicies, false)
+		if len(role.DisallowedPolicies) > 0 {
+			// We don't add the default here because we only want to disallow it if it's explicitly set
+			sanitizedRolePolicies = policyutil.SanitizePolicies(role.DisallowedPolicies, policyutil.DoNotAddDefaultPolicy)
 
-		for _, inputPolicy := range sanitizedInputPolicies {
-			if strutil.StrListContains(sanitizedRolePolicies, inputPolicy) {
-				return logical.ErrorResponse(fmt.Sprintf("token policy (%s) is disallowed by this role", inputPolicy)), logical.ErrInvalidRequest
+			for _, finalPolicy := range finalPolicies {
+				if strutil.StrListContains(sanitizedRolePolicies, finalPolicy) {
+					return logical.ErrorResponse(fmt.Sprintf("token policy (%s) is disallowed by this role", finalPolicy)), logical.ErrInvalidRequest
+				}
 			}
 		}
 
+		data.Policies = finalPolicies
+
+	// No policies specified, inherit parent
```

Comment:
Could we update the comment of all the switch cases, that they are applicable for `roles` as well? It is not intuitive, particularly for the `!isSudo` case, the comment statement ("When a role is not in use") gives a feeling that it is not applicable for roles.